packages/medusa/src/services/event-bus.ts

```diff
@@ -124,13 +148,26 @@ export default class EventBusService {
    * happens. Subscribers must return a Promise.
    * @return this
    */
-  subscribe(event: string | symbol, subscriber: Subscriber): this {
+  subscribe(
+    event: string | symbol,
+    subscriber: Subscriber,
+    context?: SubscriberContext
+  ): this {
     if (typeof subscriber !== "function") {
       throw new Error("Subscriber must be a function")
     }
 
-    const observers = this.observers_.get(event) ?? []
-    this.observers_.set(event, [...observers, subscriber])
+    const subscriberId =
+      context?.subscriberId ?? `${event.toString()}-${ulid()}`
+
+    const newSubscriberDescriptor = { subscriber, id: subscriberId }
```

Comment:
todo: this descriptor should only be added if the id does not already exists in `his.eventToSubscribersMap_.get(event)` below. Otherwise we still get the duplication. wdyt?